### PR TITLE
A function that tries to separate peaks from wells

### PR DESCRIPTION
**Pull request by eggrobin**
#3358.

### CODE REVIEW

File: physics/equipotential_body.hpp [hunk 1]
```diff
@@ -178,14 +178,151 @@ auto Equipotential<InertialFrame, Frame>::ComputeLines(
       states.push_back(State{});
       continue;
     }
-
+    LOG(ERROR) << total_energy << " "
+               << dynamic_frame_->GeometricPotential(
+                      t, equipotential_position.value());
```
Comment (pleroy) [inline physics/equipotential_body.hpp:183]: Can we remove this?

File: physics/equipotential_body.hpp [hunk 2]
```diff
@@ -178,14 +178,151 @@ auto Equipotential<InertialFrame, Frame>::ComputeLines(
...
+template<typename InertialFrame, typename Frame>
+auto Equipotential<InertialFrame, Frame>::ComputeLines(
+    Plane<Frame> const& plane,
+    Instant const& t,
+    std::vector<Position<Frame>> const& peaks,
+    std::vector<Well> const& wells,
+    std::function<Position<Frame>(Position<Frame>)> towards_infinity,
+    SpecificEnergy const& energy) const -> std::vector<State> {
+  using WellIterator = typename std::vector<Well>::const_iterator;
+  LOG(ERROR) << "V=" << energy;
```
Comment (pleroy) [inline physics/equipotential_body.hpp:200]: Do we need to log the parameter?

File: physics/equipotential_body.hpp [hunk 3]
```diff
@@ -178,14 +178,151 @@ auto Equipotential<InertialFrame, Frame>::ComputeLines(
...
+  struct PeakDelineation {
```
Comment (pleroy) [inline physics/equipotential_body.hpp:201]: Blank line before and comment.

File: physics/equipotential_test.cpp [hunk 4]
```diff
@@ -307,18 +310,83 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_EquidistantEnergies) {
...
+  KeplerianElements<Barycentric> const elements{
+      .periapsis_distance = 7000 * Kilo(Metre),
+      .apoapsis_distance = 0.8 * *moon_elements.periapsis_distance,
+      .inclination = moon_elements.inclination,
+      .longitude_of_ascending_node = moon_elements.longitude_of_ascending_node,
+      .argument_of_periapsis = *moon_elements.argument_of_periapsis + Degree,
+      .mean_anomaly = 0 * Degree};
+  auto const earth_world_dof = dynamic_frame.ToThisFrameAtTime(t0_)(earth_dof);
+  auto const moon_world_dof = dynamic_frame.ToThisFrameAtTime(t0_)(moon_dof);
+  Position<World> const qearth = earth_world_dof.position();
```
Comment (pleroy) [inline physics/equipotential_test.cpp:344]: Let's call these `q_earth`, `v_earth`, etc.

File: physics/equipotential_test.cpp [hunk 5]
```diff
@@ -307,18 +310,83 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_EquidistantEnergies) {
...
+  Position<World> const qmoon = moon_world_dof.position();
+  Velocity<World> const vearth = earth_world_dof.velocity();
+  Velocity<World> const vmoon = moon_world_dof.velocity();
+  std::vector<DegreesOfFreedom<Barycentric>> const initial_states{
+      dynamic_frame.FromThisFrameAtTime(t0_)(
+          {qmoon + 2 * (qearth - qmoon), World::unmoving}),
+      dynamic_frame.FromThisFrameAtTime(t0_)(
+          {Barycentre(std::pair(qearth, qmoon), std::pair(1, 1)) +
+               (qearth - qmoon).Norm() *
+                   Vector<double, World>({0, quantities::Sqrt(3) / 2, 0}),
```
Comment (pleroy) [inline physics/equipotential_test.cpp:354]: Give a name to this vector.

File: physics/equipotential_test.cpp [hunk 6]
```diff
@@ -307,18 +310,83 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_EquidistantEnergies) {
...
+  auto const moon_world_dof = dynamic_frame.ToThisFrameAtTime(t0_)(moon_dof);
+  Position<World> const qearth = earth_world_dof.position();
+  Position<World> const qmoon = moon_world_dof.position();
+  Velocity<World> const vearth = earth_world_dof.velocity();
+  Velocity<World> const vmoon = moon_world_dof.velocity();
+  std::vector<DegreesOfFreedom<Barycentric>> const initial_states{
+      dynamic_frame.FromThisFrameAtTime(t0_)(
+          {qmoon + 2 * (qearth - qmoon), World::unmoving}),
+      dynamic_frame.FromThisFrameAtTime(t0_)(
+          {Barycentre(std::pair(qearth, qmoon), std::pair(1, 1)) +
```
Comment (pleroy) [inline physics/equipotential_test.cpp:352]: Give a name to the calls to `Barycentre`.

File: physics/equipotential_test.cpp [hunk 7]
```diff
@@ -349,36 +417,108 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_GlobalOptimization) {
...
   for (int j = 0; j < number_of_days; ++j) {
     LOG(ERROR) << "Day #" << j;
     t = t0_ + j * Day;
     CHECK_OK(ephemeris_->Prolong(t));
     all_positions.emplace_back();
     all_βs.emplace_back();
 
-    auto const maxima = optimizer.FindGlobalMaxima(
-        /*points_per_round=*/100,
+    auto const arg_maxima = optimizer.FindGlobalMaxima(
```
Comment (pleroy) [inline physics/equipotential_test.cpp:433]: `arg_` is not a great name.  `location_maxima`, maybe?

File: physics/equipotential_test.cpp [hunk 8]
```diff
@@ -349,36 +417,108 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_GlobalOptimization) {
...
         /*number_of_rounds=*/std::nullopt,
         /*local_search_tolerance=*/10'000 * Kilo(Metre));
-    logger.Append("maxima", maxima, mathematica::ExpressIn(Metre));
-
-    SpecificEnergy maximum_energy = -Infinity<SpecificEnergy>;
-    for (auto const& maximum : maxima) {
-      maximum_energy = std::max(maximum_energy,
-                                dynamic_frame.GeometricPotential(t, maximum));
+    logger.Append("argMaxima", arg_maxima, mathematica::ExpressIn(Metre));
+    std::vector<SpecificEnergy> maxima;
```
Comment (pleroy) [inline physics/equipotential_test.cpp:438]: `maxima_energies`?

File: physics/equipotential_test.cpp [hunk 9]
```diff
@@ -307,18 +310,83 @@ TEST_F(EquipotentialTest, BodyCentredBodyDirection_EquidistantEnergies) {
...
+      .longitude_of_ascending_node = moon_elements.longitude_of_ascending_node,
+      .argument_of_periapsis = *moon_elements.argument_of_periapsis + Degree,
+      .mean_anomaly = 0 * Degree};
+  auto const earth_world_dof = dynamic_frame.ToThisFrameAtTime(t0_)(earth_dof);
+  auto const moon_world_dof = dynamic_frame.ToThisFrameAtTime(t0_)(moon_dof);
+  Position<World> const qearth = earth_world_dof.position();
+  Position<World> const qmoon = moon_world_dof.position();
+  Velocity<World> const vearth = earth_world_dof.velocity();
+  Velocity<World> const vmoon = moon_world_dof.velocity();
+  std::vector<DegreesOfFreedom<Barycentric>> const initial_states{
```
Comment (pleroy) [inline physics/equipotential_test.cpp:348]: Please comment what's going on in this test, in particular the test trajectories.

File: physics/equipotential_body.hpp [hunk 10]
```diff
@@ -178,14 +178,151 @@ auto Equipotential<InertialFrame, Frame>::ComputeLines(
...
+
+  std::vector<PeakDelineation> peak_delineations(peaks.size());
+  for (auto& delineation : peak_delineations) {
+    for (auto it = wells.begin(); it != wells.end(); ++it) {
+      delineation.indistinct_wells.insert(it);
+    }
+  }
+
+  std::vector<State> states;
+  for (int i = 0; i < peaks.size(); ++i) {
```
Comment (pleroy) [inline physics/equipotential_body.hpp:214]: This loop is going to need *a lot* more comments.

File: physics/equipotential_body.hpp (hunk 2 above)
Comment (eggrobin) [reply physics/equipotential_body.hpp:200]: Yes, otherwise the other logs are hard to interpret.

File: physics/equipotential_test.cpp (hunk 7 above)
Comment (eggrobin) [reply physics/equipotential_test.cpp:433]: arg_maximorum.

File: physics/equipotential_body.hpp [hunk 11]
```diff
@@ -198,11 +195,16 @@ auto Equipotential<InertialFrame, Frame>::ComputeLines(
     SpecificEnergy const& energy) const -> std::vector<State> {
   using WellIterator = typename std::vector<Well>::const_iterator;
   LOG(ERROR) << "V=" << energy;
+
+  // The set of wells that are not yet delineated from a peak by equipotentials
+  // already computed, as well as whether the well is delineated from the “well
```
Comment (pleroy) [inline physics/equipotential_body.hpp:200]: `as well as` -> `together with`, `also`, `plus`, `additionally`, etc.  There are two many wells in this comment. ䷯.

**Review by pleroy — APPROVED** [review]

**Pull request merged**